### PR TITLE
chore: add CI & Azure-Observe E2E testing 

### PR DESCRIPTION
**Pull request by obs-gh-nikhildua**
Overview:
- Adds a CI job to create & destroy ephemeral resource groups in the `Terraform CI` Subscription
- Opening a new PR (with branch) creates a new resource group with entire TF collection installed and also creates ephemeral Observe token in 179 environment (for now) 
- Deleting a PR (with branch) delete the entire resource group (thus removing TF collection and Observe token)
- Posts TF plan as summary and within PR
- After creating the environment, CI runs validation test to ensure data is flowing from Azure --> Observe for Resources, Metrics and FunctionApp Logs

Note: 
- All deployments are always created in `centralus` 
- Terraform state, for a given branch, is managed in a separate RG `rg-terraform-github-actions-state` with its own storage account and keys per branch 

Addresses: 
- https://observe.atlassian.net/browse/OB-27039
- https://observe.atlassian.net/browse/OB-30693
- https://observe.atlassian.net/browse/OB-29418

For more info on this PR, see READMEs:
- https://github.com/observeinc/terraform-azure-collection/blob/nikhil/add-CI-e2e-tests/.github/terraform/README.md
- https://github.com/observeinc/terraform-azure-collection/blob/nikhil/add-CI-e2e-tests/.github/scripts/README.md

Observe Token for this branch `nikhil-add-CI-e2e-tests`:
Screenshot 2024-05-02 at 4 10 19 PM

Azure Resource Group with appropriate tags:  
Screenshot 2024-05-02 at 4 10 49 PM

Diagram:
ci-tests

### CODE REVIEW

File: .github/workflows/ci-tests.yml [hunk 1]
```diff
@@ -1,7 +1,211 @@
...
+        
+    - name: Data Validation Test 
+      uses: nick-fields/retry@v3
+      if:  steps.tf-apply.outcome != 'failure' && github.event.pull_request.state != 'closed'
+      with:
+        timeout_minutes: 30
+        max_attempts: 40 #Max 40 attempts (resource data can take longer depending on when the next cron job runs)
+        retry_wait_seconds: 60  #1 attempt/minute
+        command: |
+            pip install requests && python ${{github.workspace}}/.github/scripts/query_observe.py
```
Comment (arthur-observe) [inline .github/workflows/ci-tests.yml:200]: You should manage python dependencies with a requirements.txt file

File: .github/scripts/query_observe.py [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+    # Create console handler and set level to debug
+    ch = logging.StreamHandler()
+    ch.setLevel(log_level)
+    ch.setFormatter(formatter)
+
+    # Add console handler to logger
+    logger.addHandler(ch)
+
+
+def get_bearer_token() -> str:
```
Comment (arthur-observe) [inline .github/scripts/query_observe.py:31]: Not for ths pr but we use this over an over we should have a common library and make some classes we can import so we stop repeating same code all over

File: .github/scripts/query_observe.py [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+    }
+
+    response = json.loads(
+        requests.post(url, data=message, headers=header, timeout=10).text
+    )
+    bearer_token = response['access_key']
+    return bearer_token
+
+
+def send_query(bearer_token: str, query: str, params: dict = None, url_extension: str = '',
```
Comment (arthur-observe) [inline .github/scripts/query_observe.py:62]: same  for this code on reusability

File: .github/scripts/query_observe.py [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+    current_ts_string = datetime.datetime.fromtimestamp(current_ts, timezone.utc).strftime('%Y-%m-%dT%H:%M:%SZ')
+ 
+    logger.info("{}: stale_check_mins is: {} mins".format(source, stale_checks_mins))
+    logger.info("{}: query interval is: {}".format(source, query_interval))
+    logger.info("{}: Current time is: {}".format(source, current_ts_string))
+
+    pipeline = ''
+    if source == "EventHub":
+        # OPAL PIPELINE to execute on Dataset
+        pipeline = "make_col _ob_datastream_token_id:coalesce(DATASTREAM_TOKEN_ID, string(EXTRA.datastream_token_id))| " \
```
Comment (arthur-observe) [inline .github/scripts/query_observe.py:185]: I like to store queries in their own files and import - not required but consider

File: .github/scripts/query_observe.py [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+                   "filter _ob_datastream_token_id = '{}'|" \
+                   "filter (string(EXTRA.source) = 'ResourceManagement')|" \
+                   "make_col source: string(EXTRA.source)|" \
+                   "make_col type:string(FIELDS.type)|" \
+                   "statsby msg_count: count_distinct(type), latest_ts: last_not_null(BUNDLE_TIMESTAMP), group_by(source)" \
+                   "". \
+            format(OBSERVE_TOKEN_ID)
+    elif source == "VmMetrics":
+        pipeline = "make_col _ob_datastream_token_id:coalesce(DATASTREAM_TOKEN_ID, string(EXTRA.datastream_token_id))| " \
```
Comment (arthur-observe) [inline .github/scripts/query_observe.py:211]: f string f""" """ makes multiline stuff easier

Comment (arthur-observe) [reply .github/scripts/query_observe.py:211]: also can put variables in lin {{ }}

File: .github/terraform/variables.tf [hunk 6]
```diff
@@ -0,0 +1,20 @@
+variable "branch" {
+  description = "Github Action Branch Name. This will set Observe Datastream Token Name"
```
Comment (arthur-observe) [inline .github/terraform/variables.tf:2]: missing type - linter usually bitches about this

File: .github/workflows/ci-tests.yml [hunk 7]
```diff
@@ -1,7 +1,211 @@
...
+env:
+  ARM_CLIENT_ID: "${{ secrets.AZURE_CLIENT_ID }}" # Needed for azure provider service account 
+  ARM_CLIENT_SECRET: "${{ secrets.AZURE_CLIENT_SECRET }}" # Needed for azure provider service account 
+  ARM_SUBSCRIPTION_ID: "${{ secrets.AZURE_SUBSCRIPTION_ID }}" # Needed for azure provider service account 
+  ARM_TENANT_ID: "${{ secrets.AZURE_TENANT_ID }}" # Needed for azure provider service account   
+  OBSERVE_CUSTOMER: ${{ vars.OBSERVE_CUSTOMER}} #Needed for observe.tf
+  OBSERVE_DOMAIN: ${{ vars.OBSERVE_DOMAIN}} #Needed for observe.tf
+  OBSERVE_USER_EMAIL: ${{ secrets.OBSERVE_USER_EMAIL}} #Needed for observe.tf
+  OBSERVE_USER_PASSWORD: ${{ secrets.OBSERVE_USER_PASSWORD}} #Needed for observe.tf
+  TF_VAR_observe_customer: "${{ vars.OBSERVE_CUSTOMER}}" #Fed to terraform-azure-collection
```
Comment (arthur-observe) [inline .github/workflows/ci-tests.yml:34]: almost positive these will get uppercased on you - should double check and see if tf is sensitive to case

File: .github/scripts/query_observe.py (hunk 2 above)
Comment (obs-gh-nikhildua) [reply .github/scripts/query_observe.py:31]: I agree. At least the below:
- Bearer Token
- Send Observe Query

Comment (obs-gh-nikhildua) [reply .github/scripts/query_observe.py:31]: https://observe.atlassian.net/browse/OB-31450 
For future

File: .github/scripts/query_observe.py (hunk 3 above)
Comment (obs-gh-nikhildua) [reply .github/scripts/query_observe.py:62]: https://observe.atlassian.net/browse/OB-31450
For future

File: .github/workflows/ci-tests.yml (hunk 7 above)
Comment (obs-gh-nikhildua) [reply .github/workflows/ci-tests.yml:34]: TF allows lowercase in env vars: 
https://developer.hashicorp.com/terraform/cli/config/environment-variables#tf_var_name

File: .github/workflows/ci-tests.yml (hunk 1 above)
Comment (obs-gh-nikhildua) [reply .github/workflows/ci-tests.yml:200]: Usually I would but in this case it's just one library and rest is all python native libs.

File: .github/terraform/variables.tf (hunk 6 above)
Comment (obs-gh-nikhildua) [reply .github/terraform/variables.tf:3]: Added

File: .github/scripts/query_observe.py (hunk 5 above)
Comment (obs-gh-nikhildua) [reply .github/scripts/query_observe.py]: I've tried 3 quotes & 2 quotes back-back but the \n new lines in 3 quotes don't play well with OPAL during query time as our API complains of invalid json

I decided to separate out the OPAL into its own file as you mentioned to make it easy to read

File: .github/scripts/query_observe.py (hunk 4 above)
Comment (obs-gh-nikhildua) [reply .github/scripts/query_observe.py]: Done! thx

**Review by arthur-observe — APPROVED** [review]
Unresolved the case comment and put a link to issue that is similar - I think GHA will uppercase your env vars no matter what u you do but it's not obvious unless you print them out.  So you should be really sure the values you expect when you set them outside defaults happen.

**Comment by obs-gh-nikhildua** [comment]
> Unresolved the case comment and put a link to issue that is similar - I think GHA will uppercase your env vars no matter what u you do but it's not obvious unless you print them out. So you should be really sure the values you expect when you set them outside defaults happen.
Screenshot 2024-05-14 at 6 25 42 PM


@arthur-observe Are you referring to these? Haven't been able to replicate it

**Pull request merged**